File: 1282. Group the People Given the Group Size They Belong To.rb
#*1282. Group the People Given the Group Size They Belong To
#*LEARN => zip group_by map transpose each_slice hash to_a array
# @param {Integer[]} group_sizes
# @return {Integer[][]}
#* HELL YES
def group_the_people(group_sizes = [3, 3, 3, 3, 3, 1, 3])
  group_sizes.zip(0..).group_by(&:first).map{|group,indices| indices.transpose.last.each_slice(group).to_a}.sum([])
end
p group_the_people

#*2
def group_the_people(group_sizes = [3, 3, 3, 3, 3, 1, 3])
  groups = group_sizes.each_with_index.group_by {|value, index| value}.transform_values { |value| value.map(&:last) }
  result = []
  groups.each {|k, v| v.each_slice(k) {|a| result << a}}
  result
end
p group_the_people

#* LEARN => ZIP
arr=[1,2,3]
array=[11,12,13]
p arr.zip(array)

names = ["Mary", "John", "Emma"]
heights = [180, 165, 170]

p names.zip(heights).sort_by{ |k,v| -v}

File: 1282. Group the People Given the Group Size They Belong To.js
//1282. Group the People Given the Group Size They Belong To
/**
 * @param {number[]} groupSizes //<= group
 * @return {number[][]}
 */
var groupThePeople = function(group = [2,1,3,3,3,2]) {
  const res = []
  const hash = {}
    
  for (let i = 0; i < group.length; i++) {
    const num = group[i]
    hash[num] ? hash[num].push(i) : hash[num] = [i]
    
    if (hash[num].length === num) {
      res.push(hash[num])
      delete hash[num]
    }
  }
  return res
};
console.log(groupThePeople())